scripts/apps/master-desk/MasterDesk.tsx

```diff
@@ -0,0 +1,183 @@
+import React from 'react';
+import {IDesk, IStage, IUser} from 'superdesk-api';
+
+import {
+    HeaderComponent,
+    OverviewComponent,
+    UsersComponent,
+    AssignmentsComponent,
+    PreviewComponent,
+    FilterPanelComponent,
+    FilterBarComponent,
+} from 'apps/master-desk/components';
+
+import {assertNever} from 'core/helpers/typescript-helpers';
+import {gettext} from 'core/utils';
+import {appConfig} from 'appConfig';
+
+import UserActivityWidget from 'apps/dashboard/user-activity/components/UserActivityWidget';
+
+export enum IMasterDeskTab {
+    overview = 'overview',
+    users = 'users',
+    assignments = 'assignments',
+}
+
+export const USER_PREFERENCE_SETTINGS = 'masterdesk:desks';
+
+export function getLabelForMasterDeskTab(tab: IMasterDeskTab): string {
+    switch (tab) {
+    case IMasterDeskTab.overview:
+        return gettext('Overview');
+    case IMasterDeskTab.users:
+        return gettext('Users');
+    case IMasterDeskTab.assignments:
+        return gettext('Assignments');
+    default:
+        return assertNever(tab);
+    }
+}
+
+interface IProps {
+    desks: any;
+    api: any;
+    tasks: any;
+    preferencesService: any;
+}
+
+interface IState {
+    desks: Array<IDesk>;
+    stages: Array<IStage>;
+    currentTab: IMasterDeskTab;
+    filterOpen: boolean;
+    activeUser: IUser;
+    planning: boolean;
+    deskFilter: string;
+    filters: any;
+}
+
+export class MasterDesk extends React.Component<IProps, IState> {
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {
+            currentTab: IMasterDeskTab.overview,
+            filterOpen: false,
+            desks: [],
+            stages: [],
+            activeUser: null,
+            planning: false,
+            deskFilter: '',
+            filters: {},
+        };
+    }
+
+    componentDidMount() {
+        this.props.preferencesService.get(USER_PREFERENCE_SETTINGS).then((desks) => {
+            !desks || !desks.items.length ?
+                this.getDeskList() :
+                this.getDeskList(desks.items);
+        });
+
+        if (appConfig?.apps?.includes('superdesk-planning')) {
+            this.setState({planning: true});
+        }
+    }
+
+    getDeskList(enabledDeskIds?: Array<string>) {
+        const desks = this.props.desks;
+
+        desks.initialize().then(() => {
+            this.setState({stages: desks.deskStages});
+
+            let filteredDesks;
+
+            enabledDeskIds && enabledDeskIds.length ?
+                filteredDesks = desks.desks._items.filter((item) => enabledDeskIds.includes(item._id)) :
+                filteredDesks = desks.desks._items;
+
+            this.setState({desks: filteredDesks});
+        });
+    }
+
+    isFilterAllowed() {
+        return this.state.currentTab === IMasterDeskTab.overview;
+    }
+
+    render() {
```

Comment:
In feature plan is to use filter component also on assignments and on users. Currently functionality is only developed for Overview panel, but soon it will be prepared for all 3.

Reply:
It would be good to at least move active user preview to users page then. To do that, you would probably still need to move the entire layout to each separate page. You could then pass filters as a prop in the same way as I suggested to pass `headerComponent`.

Reply:
I can't move it to `UsersComponent` because of layout. I need this panel on top of html structure, so it can look and behave nice. Also, there is a plan in future to use this `PreviewComponent` on other pages, and because of that I created wrapper `PreviewComponent` and will use it for other components.

Reply:
You might use preview panel in other pages, but **user** preview is specific to the user page.

> I can't move it to `UsersComponent` because of layout.

Yes you can :) You just need to use the root layout in each page and pass common things as props:

```
class UsersComponent extends React.Component {
    render() {
        return (
            <div className="sd-content-wrapper__main-content-area sd-main-content-grid">
                {header}

                {filteringPane}

                <div className="sd-main-content-grid__content">
                    {selectedFiltersBar}

                    <div className="sd-main-content-grid__content-inner">
                        <UsersComponent
                            desks={this.state.desks}
                            deskService={this.services.desks}
                            onUserSelect={(user) => this.setState({activeUser: user})}
                        />
                    </div>
                </div>

                {this.state.activeUser ? (
                    <PreviewComponent header={'User Activity'} onClose={() => this.setState({activeUser: null})}>
                        <UserActivityWidget user={this.state.activeUser} />
                    </PreviewComponent>
                ) : null}
            </div>
        );
    }
}
```

Reply:
I prefer from UI point of view not to duplicate layout structure 🙂  Let's keep it like this for sake of consistence, please. Thanks

Reply:
What would be inconsistent in my suggestion?

You can create a common master page layout component if you think it helps not to duplicate layout structure.

There doesn't have to be a trade-off here - you can have both - user component in it's own page without duplicating the layout.

```
interface IProps {
    header: JSX.Element;
    content: JSX.Element;
    filteringBar: JSX.Element;
    sidebar: JSX.Element;
}

class MasterDeskPage extends React.Component<IProps> {
    render() {
        const {header, content, filteringBar, sidebar} = this.props;

        return (
            <div className="sd-content-wrapper__main-content-area sd-main-content-grid">
                {header}

                <div className="sd-main-content-grid__content">
                    {filteringBar}

                    <div className="sd-main-content-grid__content-inner">
                        {content}
                    </div>
                </div>

                {sidebar}
            </div>
        );
    }
}

// -----

class UsersComponent extends React.Component {
    render() {
        return (
            <MasterDeskPage
                header={(
                    <React.Fragment>
                        {header}
                        {filteringPane}
                    </React.Fragment>
                )}
                filteringBar={selectedFiltersBar}
                content={(
                    <UsersComponent
                        desks={this.state.desks}
                        deskService={this.services.desks}
                        onUserSelect={(user) => this.setState({activeUser: user})}
                    />
                )}
                sidebar={this.state.activeUser ? (
                    <PreviewComponent header={'User Activity'} onClose={() => this.setState({activeUser: null})}>
                        <UserActivityWidget user={this.state.activeUser} />
                    </PreviewComponent>
                ) : null}
            />
        );
    }
}
```

Reply:
If you are busy with UI framework I can help to adjust this since there aren't too many client core tasks this sprint.

Reply:
Let's see to merge it like this, so we can deploy it to 2.1, not to postpone it anymore, and we can do some changes after for some feature releases.

Reply:
No problem. Can you then submit this fix as a separate PR? In the meantime I will check your latest commit and will approve the PR if I don't see any significant issues.